modules/ui/src/main/scala/lucuma/ui/forms/FormInputEV.scala

```diff
@@ -99,35 +108,101 @@ object FormInputEV {
+        .flatMap(oc => $.setStateL(State.cursor)(oc))
+
+    def clearStateCursor: Callback = $.setStateL(State.cursor)(None)
+
+    def setCursorFromState: Callback =
+      $.state.flatMap(s => s.cursor.map(setCursor).getOrElse(Callback.empty))
+
+    def audit(auditor: ChangeAuditor[A], value: String): CallbackTo[String] = {
+      def setDisplayValue(s: String): CallbackTo[String] =
+        $.setStateL(State.displayValue)(s).map(_ => s)
+
+      def cursorOffsetForReject: CallbackTo[Int] =
+        $.state.map(_.lastKeyCode match {
+          case KeyCode.Backspace => 1
+          case KeyCode.Delete    => 0
+          case _                 => -1
+        })
+
+      getCursor
+        .map {
+          case Some(c) => c._1
+          case _       => value.length
+        }
+        .flatMap { c =>
+          auditor.audit(value, c) match {
+            case AuditResult.Accept                  => clearStateCursor *> setDisplayValue(value)
+            case AuditResult.NewString(newS, offset) =>
+              setStateCursor(offset) *> setDisplayValue(newS)
+            case AuditResult.Reject                  =>
+              cursorOffsetForReject.flatMap(setStateCursor _) *> CallbackTo(value)
+          }
+        }
+    }
+
     def onTextChange(props: Props[EV, A]): ReactEventFromInput => Callback =
       (e: ReactEventFromInput) => {
         // Capture the value outside setState, react reuses the events
         val v = e.target.value
-        // First update the internal state, then call the outside listener
-        $.setStateL(State.curValue)(v) *> $.setStateL(State.errors)(none) *>
-          props.onTextChange(v) *>
-          validate(props, v)
+        audit(props.changeAuditor, v).flatMap(newS =>
+          // First update the internal state, then call the outside listener
+          $.setStateL(State.errors)(none) *> props.onTextChange(newS) *> validate(props, newS)
+        )
       }
 
     def onBlur(props: Props[EV, A], state: State): Callback =
       validate(
         props,
-        state.curValue,
+        state.displayValue,
         { validated =>
           val validatedCB = validated match {
             case Valid(a)   =>
               implicit val eq = props.eq
               if (props.ev.get(props.value).exists(_ =!= a)) // Only set if resulting A changed.
                 props.valSet(a)
               else                                           // A didn't change, but redisplay formatted string.
-                $.setStateL(State.curValue)(props.valGet)
+                $.setStateL(State.displayValue)(props.valGet)
             case Invalid(e) =>
               $.setStateL(State.errors)(e.some)
           }
           validatedCB >> props.onBlur(validated)
         }
       )
 
+    val onKeyDown: ReactKeyboardEventFromInput => Callback = e =>
+      $.setStateL(State.lastKeyCode)(e.keyCode) *> clearStateCursor
+
+    val OuterDiv = <.div()
```

Comment:
Can we change this into a `span` so that it doesn't force a `block` display? Also, I think you don't need the `()`.

Reply:
Good idea.